desc/optimize/_constraint_wrappers.py

```diff
@@ -846,35 +858,118 @@ def hess(self, x, constants=None):
 
-    def vjp_scaled(self, v, x):
-        """Compute vector-Jacobian product of the objective function.
+    def jvp_scaled(self, v, x, constants=None):
+        """Compute Jacobian-vector product of the objective function.
 
         Uses the scaled form of the objective.
 
         Parameters
         ----------
-        v : ndarray
-            Vector to left-multiply the Jacobian by.
+        v : ndarray or tuple of ndarray
+            Vectors to right-multiply the Jacobian by.
+            This method only works for first order jvps.
         x : ndarray
             Optimization variables.
+        constants : list
+            Constant parameters passed to sub-objectives.
 
         """
-        raise NotImplementedError
+        v = v[0] if isinstance(v, (tuple, list)) else v
+        constants = setdefault(constants, self.constants)
+        xg, xf = self._update_equilibrium(x, store=True)
+        jvpfun = lambda u: self._jvp(u, xf, xg, constants, op="scaled")
+        return jnp.vectorize(jvpfun, signature="(n)->(k)")(v)
 
-    def vjp_unscaled(self, v, x):
-        """Compute vector-Jacobian product of the objective function.
+    def jvp_unscaled(self, v, x, constants=None):
+        """Compute Jacobian-vector product of the objective function.
 
         Uses the unscaled form of the objective.
 
         Parameters
         ----------
-        v : ndarray
-            Vector to left-multiply the Jacobian by.
+        v : ndarray or tuple of ndarray
+            Vectors to right-multiply the Jacobian by.
+            This method only works for first order jvps.
         x : ndarray
             Optimization variables.
+        constants : list
+            Constant parameters passed to sub-objectives.
 
         """
-        raise NotImplementedError
+        v = v[0] if isinstance(v, (tuple, list)) else v
+        constants = setdefault(constants, self.constants)
+        xg, xf = self._update_equilibrium(x, store=True)
+        jvpfun = lambda u: self._jvp(u, xf, xg, constants, op="unscaled")
+        return jnp.vectorize(jvpfun, signature="(n)->(k)")(v)
+
+    def _jvp_f(self, xf, dc, constants, op):
+        Fx = getattr(self._constraint, "jac_" + op)(xf, constants)
+        Fx_reduced = Fx[:, self._unfixed_idx] @ self._Z
+        Fc = Fx @ (self._dxdc @ dc)
+        Fxh = Fx_reduced
+        cutoff = jnp.finfo(Fxh.dtype).eps * max(Fxh.shape)
+        uf, sf, vtf = jnp.linalg.svd(Fxh, full_matrices=False)
+        sf += sf[-1]  # add a tiny bit of regularization
+        sfi = jnp.where(sf < cutoff * sf[0], 0, 1 / sf)
+        Fxh_inv = vtf.T @ (sfi[..., None] * uf.T)
+        return Fxh_inv @ Fc
+
+    def _jvp(self, v, xf, xg, constants, op):
+        # we're replacing stuff like this with jvps
+        # Fx_reduced = Fx[:, unfixed_idx] @ Z               # noqa: E800
+        # Gx_reduced = Gx[:, unfixed_idx] @ Z               # noqa: E800
+        # Fc = Fx @ dxdc @ v                                # noqa: E800
```

Comment:
What is the order of products here? Maybe put parentheses?

Reply:
matmul is associative so its whatever order the compiler decides